arm64: dts: qcom: msm8916-longcheer-l8910 + msm8939-longcheer-l9100: Enable RGB LED

a-andre opened the pull request.

Commit:
arm64: dts: qcom: msm8939-longcheer-l9100: Enable RGB LED

l9100 uses KTD2026 LED driver. Add it to the device tree.

Signed-off-by: André Apitzsch <git@apitzsch.eu>

TravMurav commented on drivers/leds/leds-ktd202x.c:
```diff
@@ -0,0 +1,417 @@
+// SPDX-License-Identifier: GPL-2.0+
```
```suggestion
// SPDX-License-Identifier: GPL-2.0-or-later
```

(see https://spdx.org/licenses/#deprecated)

TravMurav commented on drivers/leds/leds-ktd202x.c:
```diff
@@ -0,0 +1,417 @@
+	chip->vcc_regulator = devm_regulator_get(&client->dev, "vcc");
+	ret = PTR_ERR_OR_ZERO(chip->vcc_regulator);
+	if (ret) {
+		if (ret != -EPROBE_DEFER)
+			dev_err(&client->dev,
+				"Failed to request regulator: %d\n", ret);
```
there is amazing `dev_err_probe()` you could use here I think, it would nicely save the deferred probe reason in the sysfs devices_deferred

 Would be nice to convert all error paths to using it since it decodes the errno to the enum name as well with a special `%pe` format, so you will get i.e. `error -EINVAL: Failed to do whatever`

TravMurav commented on drivers/leds/leds-ktd202x.c:
```diff
@@ -0,0 +1,417 @@
+
+MODULE_AUTHOR("André Apitzsch <git@apitzsch.eu>");
+MODULE_DESCRIPTION("Kinetic KTD2026/7 LED driver");
+MODULE_LICENSE("GPL v2");
```
iirc this should be just GPL too, checkpatch should say

TravMurav commented on Documentation/devicetree/bindings/leds/leds-ktd202x.yaml:
```diff
@@ -0,0 +1,89 @@
+    unevaluatedProperties: false
+
+    properties:
+      reg:
```
might want to mark reg as required for the subnode

TravMurav commented on Documentation/devicetree/bindings/leds/leds-ktd202x.yaml:
```diff
@@ -0,0 +1,89 @@
+
+examples:
+  - |
+    #include <dt-bindings/gpio/gpio.h>  // FIXME: really needed?
```
```suggestion
```

there are no defines from it in the example so no.

I suppose you should also run the bindings checker on this file if you didn't yet.

a-andre replied:
No errors with `make dt_binding_check DT_SCHEMA_FILES=leds-ktd202x.yaml`

TravMurav commented on arch/arm64/boot/dts/qcom/msm8939-longcheer-l9100.dts:
```diff
@@ -126,6 +126,37 @@
+
+		led@0 {
+			reg = <0>;
+			function = LED_FUNCTION_STATUS;
```
```suggestion
			function = LED_FUNCTION_INDICATOR;
```

I think we use that mostly

a-andre replied:
Because of [`/* Miscelleaus functions. Use functions above if you can. */`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/dt-bindings/leds/common.h#n75) I used `LED_FUNCTION_STATUS`. But yaml still contains `LED_FUNCTION_INDICATOR`, so any of these should be updated.

Is there any functional difference between `LED_FUNCTION_STATUS` and `LED_FUNCTION_INDICATOR`? I couldn't find anything.

TravMurav replied:
The only difference is how userspace decides to interpret it. I suppose both status and indicator is fine, i.e. default feedbackd udev rule catches both.

https://source.puri.sm/Librem5/feedbackd/-/blob/main/debian/feedbackd.udev#L10-11

So you don't need to change it if you want to use status, just that I have an impression most things apart from L5 use indicator

(Edit) Actually no idea why we used indicator in so many places given "status" is clearly defined to serve this purpose....

TravMurav reviewed (commented):
Feeling a de-ja-vu, I guess this is based on aw2013 xD

TravMurav commented on drivers/leds/leds-ktd202x.c:
```diff
@@ -0,0 +1,416 @@
+
+		ret = of_property_read_u32(child, "reg", &source);
+		if (ret != 0 || source >= ktd202x->num_leds) {
+			dev_err_probe(&chip->client->dev, ret,
```
I think here you'd keep a normal message since this doesn't actually fail probe but also have some generic "failed to parse the dt" in probe after call to probe_dt()

TravMurav commented on drivers/leds/leds-ktd202x.c:
```diff
@@ -0,0 +1,416 @@
+
+	chip->regmap = devm_regmap_init_i2c(client, &ktd202x_regmap_config);
+	if (IS_ERR(chip->regmap)) {
+		ret = PTR_ERR(chip->regmap);
+		dev_err_probe(&client->dev, ret,
+			      "Failed to allocate register map.\n");
```
if you'd think it would be more pretty this way, 
```suggestion
		ret = dev_err_probe(&client->dev, PTR_ERR(chip->regmap),
			      "Failed to allocate register map.\n");
```

TravMurav commented on drivers/leds/leds-ktd202x.c:
```diff
@@ -0,0 +1,416 @@
+
+	regmap_write(chip->regmap, KTD202X_REG_RESET_CONTROL,
+		     KTD202X_RSTR_RESET);
+	udelay(200); /* allow the device to execute the complete reset */
```
Checkpatch has some thoughts (see check in CI)

a-andre replied:
Any recommendations for a good range if `usleep_range` should be applied here? [Datasheet (p. 13)](https://www.kinet-ic.com/uploads/KTD2026-7-04h.pdf) just says 200µs.

TravMurav replied:
latest time I saw it in a driver, it was 10us, tho I'd be generous and give extra 100us here just so the scheduler has more room, it doesn't hurt to sleep more here.

Commit:
arm64: dts: qcom: msm8916-longcheer-l8910: Enable RGB LED

l8910 uses KTD2026 LED driver. Add it to the device tree.

Signed-off-by: André Apitzsch <git@apitzsch.eu>

a-andre commented:
I also have added the LED to [BQ X5 as it seems to have the same RGB LED](https://github.com/msm8916-mainline/linux-downstream/blob/bq-picmt/arch/arm/boot/dts/qcom/picmt.dts#L69-L78). But I cannot test it.

stephan-gh commented:
Works on X5. Thanks! I have some minor comments about the behavior:

- The red LED is still forced on when connecting USB/charger, even if sysfs says the red LED is off. I wonder if `qcom,led-ctrl-gpio = <&pm8916_gpios 3 0>;` is used for this and we should integrate it somehow?
- For some reason the configuration is preserved after reboots. If I enable some of the LEDs (e.g. green and blue) they turn off briefly during reboot but are then enabled again while in lk2nd. Same applies for power off, then they turn on immediately again after pressing the power button. Maybe the chip should be reset in a `.shutdown` hook?

a-andre commented:
> Works on X5. Thanks! I have some minor comments about the behavior:
> 
>     * The red LED is still forced on when connecting USB/charger, even if sysfs says the red LED is off. I wonder if `qcom,led-ctrl-gpio = <&pm8916_gpios 3 0>;` is used for this and we should integrate it somehow?
> 
>     * For some reason the configuration is preserved after reboots. If I enable some of the LEDs (e.g. green and blue) they turn off briefly during reboot but are then enabled again while in lk2nd. Same applies for power off, then they turn on immediately again after pressing the power button. Maybe the chip should be reset in a `.shutdown` hook?

Both should be fixed now, though I'm not quite happy with the documentation of the ctrl gpio.

stephan-gh commented on drivers/leds/leds-ktd202x.c:
```diff
@@ -0,0 +1,436 @@
+		return dev_err_probe(&chip->client->dev, PTR_ERR(ctrl_gpio),
+				     "Failed to get ctrl gpio.\n");
+
+	gpiod_direction_output(ctrl_gpio, 1);
```
I don't think this is really specific to the ktd202x, it's probably built around it on the mainboard somehow. Can you do this only with pinctrl? Like here, note the output-high.

https://github.com/msm8916-mainline/linux/blob/313721a30c059b4c5c9495c046157a3dbda52d2e/arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts#L420-L426

a-andre replied:
Done.

stephan-gh commented on Documentation/devicetree/bindings/leds/leds-ktd202x.yaml:
```diff
@@ -0,0 +1,93 @@
+    maxItems: 1
+
+  vcc-supply:
+    description: Regulator providing power to the "VCC" pin.
```
Does a VCC pin exists? Datasheet suggests it's called "VIN"

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-longcheer-l8910.dts:
```diff
@@ -105,6 +105,40 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		vcc-supply = <&pm8916_l6>;
```
I think the relevant vin-supply would be pm8916_l17 (because it has the 2.7V to 5.5V mentioned in the KTD datasheet).

pm8916_l6 is more complicated. It's probably needed for the i2c I/O lines to work, but it's not directly part of the ktd2026. We had some discussion about this for aw2013 in https://lore.kernel.org/linux-leds/922eab51-6931-8533-db51-51cd911a36b3@linaro.org/
AFAIU the reply from Rob it's okay-ish to add a vddio-supply for this additionally.

stephan-gh commented on Documentation/devicetree/bindings/leds/leds-ktd202x.yaml:
```diff
@@ -0,0 +1,93 @@
+
+            vcc-supply = <&pm8916_l6>;
+
+            led@0 {
```
I'm not sure how strict upstream is on this for new drivers, but you might get bonus points if you can get the multiled stuff working here for exposing the 3 LEDs as a single (RGB) LED, like in https://github.com/torvalds/linux/blob/d635f6cc934bcd467c5d67148ece74632fd96abf/Documentation/devicetree/bindings/leds/leds-qcom-lpg.yaml#L160-L180

aw2013 doesn't implement this though so you would need to look for other examples.

a-andre replied:
Following the mailing list upstream doesn't seem to request multi-led for new drivers. I implemented in anyway.

With multi-led supported implemented, I wonder if single-led supported is still needed/useful.

TravMurav replied:
I think it's always possible that some vendor placed one of these chips to drive dedicated leds in the housing. Consider a router-like box baked by such chip:
```
+---------------------------+
|  o         o        o     |
| wifi      wan      lan    |
+---------------------------+
```

then it would be weird to model it as an rgb led

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-longcheer-l8910.dts:
```diff
@@ -516,6 +550,13 @@
+&pm8916_gpios {
+	status_led_default: status-led-default-state {
+		pins = "gpio3";
+		output-high;
```
It's usually good practice to fully configure pinctrl, i.e. also specify an explicit `function`, `bias-xyz` and for pm8916_gpios also a `power-source` (see e.g. disp_en_default in osprey).

It's a bit tricky here since downstream doesn't seem to specify this properly either...

Looks like it just uses the default of PM8916_GPIO_VPH for power-source

a-andre commented:
Anything left to do before sending upstream?

stephan-gh commented:
Seems like there was some review feedback upstream :)

wonderfulShrineMaidenOfParadise commented:
#324 is opened for my KTD2026 changes.

Commit:
dt-bindings: leds: Add Kinetic KTD2026/2027 LED

Document Kinetic KTD2026/2027 LED driver devicetree bindings.

Reviewed-by: Krzysztof Kozlowski <krzysztof.kozlowski@linaro.org>
Link: https://lore.kernel.org/r/20231002-ktd202x-v6-1-26be8eefeb88@apitzsch.eu
Signed-off-by: André Apitzsch <git@apitzsch.eu>

Commit:
leds: add ktd202x driver

This commit adds support for Kinetic KTD2026/7 RGB/White LED driver.

Link: https://lore.kernel.org/r/20231002-ktd202x-v6-2-26be8eefeb88@apitzsch.eu
Signed-off-by: André Apitzsch <git@apitzsch.eu>

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-longcheer-l8910.dts:
```diff
@@ -120,6 +121,45 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		vin-supply = <&pm8916_l17>;
```
`vio-supply = <&pm8916_l6>;`

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8939-longcheer-l9100.dts:
```diff
@@ -126,6 +127,45 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		vin-supply = <&pm8916_l17>;
```
same

stephan-gh reviewed (commented):
Not sure about the open comment from TravMurav on the DT changes (LED_FUNCTION_STATUS vs LED_FUNCTION_INDICATOR).

Can you send the DT changes upstream too? Looks like the driver was applied upstream. Thanks a lot for going through the effort to upstream it!

wonderfulShrineMaidenOfParadise commented on Documentation/devicetree/bindings/leds/leds-ktd202x.yaml:
```diff
@@ -0,0 +1,164 @@
+        const: 0
+
+    patternProperties:
+      "^led@[0-3]$":
```
Does `led@3` exist?

a-andre replied:
ktd2027 has 4 leds

wonderfulShrineMaidenOfParadise replied:
What is the color of 4th led? I have an additional red led not under control of slot 1, 2 or 3.

a-andre replied:
The 4th led is white, according to the data sheet https://www.kinet-ic.com/uploads/KTD2026-7-04h.pdf

TravMurav replied:
I'd imagine the board vendor could decide to connect any leds to the controller, since they are not part of the package though, i.e. 4 dedicated green leds or an rgb led package + some other led...

a-andre commented:
> Can you send the DT changes upstream too?

Yes.

@stephan-gh Have you tested if it still works with l8910? Can I add your tested-by?

stephan-gh commented:
Can you rebase this PR with the driver version that landed upstream on `msm8916/6.6-rc5`? Use

```
$ b4 shazam -tl 20231002-ktd202x-v6-0-26be8eefeb88@apitzsch.eu
```

Also please add the DT changes I suggested. Then I'll re-test it on l8910. :)

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-longcheer-l8910.dts:
```diff
@@ -109,6 +110,45 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		vin-supply = <&pm8916_l6>;
```
Oh no, I meant adding vio-supply = <&pm8916_l6>; additionally. The vin-supply here (l17) was correct :sweat_smile:

stephan-gh commented:
Works fine on L8910. :+1:

stephan-gh commented:
Integrated into `msm8916/6.6-rc6`, thanks!
